### PR TITLE
make CompoundIntentMatcher merge the scores when both matchers fail

### PR DESCRIPTION
**Pull request by wilmol**
Hi @rabidgremlin have had a fun bug :)

Basically if both intent matchers fail to match, Compound matcher only returns the scores of the second matcher.

This is a problem because we have a machine learning matcher, then a templated matcher, so the templated matcher was clearing the scores of the machine learning matcher; which broke our label suggester in Concord.

Wasn't an issue prior to v6 because the TemplatedIntentMatcher never use to set any scores in  the debug values (so never overrid any scores).

Have fixed by putting the machine learning matcher last; but putting a fix here which merges the scores.

### CODE REVIEW

File: mutters-core/src/main/java/com/rabidgremlin/mutters/core/MatcherScores.java [hunk 1]
```diff
@@ -42,37 +47,32 @@ public MatcherScores(SortedMap<Double, SortedSet<String>> scores)
...
    * Add a score for an intent to the matcher scores.
-   * 
+   *
    * @param intent The name of the intent to add the score for.
    * @param score  The score for the intent.
    * @return returns this.
    */
   public MatcherScores addScore(String intent, Double score)
   {
+    // TODO what if there are duplicate scores for some intent?
```
Comment (rabidgremlin) [inline mutters-core/src/main/java/com/rabidgremlin/mutters/core/MatcherScores.java]: Yeah, this is probably the biggest problem with the change. What do we do with the dups ???

File: mutters-core/src/main/java/com/rabidgremlin/mutters/core/MatcherScores.java [hunk 2]
```diff
@@ -97,4 +97,67 @@ public OptionalDouble getBestScore()
...
+      return secondScores;
+    }
+    else if (secondScores.isEmpty())
+    {
+      return firstScores;
+    }
+
+    // flatten [Double, Set[String]] into [String, Double]
+    // taking the highest score for any duplicate intent scores
+    Map<String, Double> flattenedScores = Stream
```
Comment (rabidgremlin) [inline mutters-core/src/main/java/com/rabidgremlin/mutters/core/MatcherScores.java]: Taking the highest score works but will skew the scoring. Also one matcher may be wildly out in its prediction so highest score is not necessary a good score :(

**Comment by rabidgremlin** [comment]
Maybe better approach is for consumer of CompoundIntentMatcher to specify which matcher is the primary "scorer" and that matcher's scores are used in a "no match scenario" ? Would give the consumer the control and make it obvious what is going to happen ?

**Comment by wilmol** [comment]
@rabidgremlin 

for my case, one scorer always has empty scores, so combining the scores doesnt create any problems.

I think you're right, maybe the user should provide their own strategy to merge the scores. E.g. constructor takes `BinaryOperator<MatcherScores> noScoreMerger` and invoke this if two no matches.

But then concerned about modifying constructor signature....

**Comment by wilmol** [comment]
Maybe could dump all the scores into a `MatcherScores` and user can just clean up afterwards.

**Comment by rabidgremlin** [comment]
We could have a version of the constructor as is and have it implement a default strategy for merging which can probably be "return no scores" and then detail in the docs.

**Comment by rabidgremlin** [comment]
Have changed the base branch to v7...

**Pull request merged**